File: LintcodePartI/li127_topologicalSorting.py

# Definition for a Directed graph node
# class DirectedGraphNode:
#     def __init__(self, x):
#         self.label = x
#         self.neighbors = []

class MyQueue:
    
    def __init__(self):
        self.item = []
    def enqueue(self, x):
        self.item.insert(0,x)
    def dequeue(self):
        return self.item.pop()
    def empty(self):
        return len(self.item) == 0
    def size(self):
        return len(self.item)
        
        
class Solution:  # TLE problem, I suggest to use a personal implemented queue class
    """
    @param graph: A list of Directed graph node
    @return: A list of graph nodes in topological order.
    """
    def topSort(self, graph):
        # write your code here
        hash = {}
        q = MyQueue()
        result = []
        for nd in graph:
            for x in nd.neighbors:
                if x not in hash.keys():
                    hash[x] = 1
                else:
                    hash[x] += 1

        for nd in graph:
            if nd not in hash.keys():
                q.enqueue(nd)
                result.append(nd)
        
        while not q.empty():
            head = q.dequeue()
            for nbh in head.neighbors:
                hash[nbh] -= 1
                if hash[nbh] == 0:
                    q.enqueue(nbh)
                    result.append(nbh)
                
        return result
    
    
import Queue
class Solution2:  # this also has TLE problem
    """
    @param graph: A list of Directed graph node
    @return: A list of graph nodes in topological order.
    """
    def topSort(self, graph):
        # write your code here
        hash = {}
        q = Queue.Queue(maxsize = len(graph))
        result = []
        for nd in graph:
            for x in nd.neighbors:
                if x not in hash.keys():
                    hash[x] = 1
                else:
                    hash[x] += 1

        for nd in graph:
            if nd not in hash.keys():
                q.put(nd)
                result.append(nd)
        
        while not q.empty():
            head = q.get()
            for nbh in head.neighbors:
                hash[nbh] -= 1
                if hash[nbh] == 0:
                    q.put(nbh)
                    result.append(nbh)
                
        return result

File: LintcodePartI/li127_topologicalSorting.java

/*Given an directed graph, a topological order of the graph nodes is defined as follow:

For each directed edge A -> B in graph, A must before B in the order list.
The first node in the order can be any node in the graph with no nodes direct to it.
Find any topological order for the given graph.

 Notice

You can assume that there is at least one topological order in the graph.

Have you met this question in a real interview? Yes
Clarification
Learn more about representation of graphs

Example
For graph as follow:

picture

The topological order can be:

[0, 1, 2, 3, 4, 5]
[0, 2, 3, 1, 5, 4]
...
*/
/**
 * Definition for Directed graph.
 * class DirectedGraphNode {
 *     int label;
 *     ArrayList<DirectedGraphNode> neighbors;
 *     DirectedGraphNode(int x) { label = x; neighbors = new ArrayList<DirectedGraphNode>(); }
 * };
 */
public class Solution {
    /**
     * @param graph: A list of Directed graph node
     * @return: Any topological order for the given graph.
     */    
    public ArrayList<DirectedGraphNode> topSort(ArrayList<DirectedGraphNode> graph) {
        // write your code here
        ArrayList<DirectedGraphNode> result = new ArrayList<DirectedGraphNode>();
        Queue<DirectedGraphNode> queue = new  LinkedList<DirectedGraphNode>();
        HashMap<DirectedGraphNode, Integer> hash =  new HashMap<DirectedGraphNode, Integer>();
        // find all of the nodes' in degrees (except for 0 indegree)
        for (DirectedGraphNode dn: graph) {
            for (DirectedGraphNode x: dn.neighbors) {
                if (hash.containsKey(x)) {
                    hash.put(x, hash.get(x) + 1);
                } else {
                    hash.put(x, 1);
                }
            }
        }
        // push nodes in the queue for in degree = 0
        for (DirectedGraphNode nd: graph) {
            if (!hash.containsKey(nd)) {
                queue.offer(nd);
                result.add(nd);
            }
        }
        // update the indegrees by removing nodes
        
        while (!queue.isEmpty()) {
            DirectedGraphNode head = queue.poll();
            for (DirectedGraphNode nd: head.neighbors) {
                hash.put(nd, hash.get(nd) - 1);
                if (hash.get(nd) == 0) {
                    queue.offer(nd);
                    result.add(nd);
                }
            }
        }
        return result;
        
    }
}


